### DISCUSSION TITLE
Open email containing html attachment with qutebrowser

### DISCUSSION THREAD

**Discussion by wgmayer0** [neomutt/neomutt#4508] [Development]
Is there a way to open Neomutt attachments automatically? For example, here are my emails:

![1640741592](https://user-images.githubusercontent.com/73801421/147619123-1ed6e926-194c-4308-9122-78520f2c33ae.png)

And here is an attachment view of the first email:

![1640741689](https://user-images.githubusercontent.com/73801421/147619216-27fa3dcf-6a2e-4a0c-8227-384108636204.png)

As you can see, one of the attachments has the extension `text/html`.

I would like Neomutt to automatically open any and all html files in Qutebrowser if they exist. (https://github.com/qutebrowser/qutebrowser) If they don't exist, I just want it to open whatever text is associated with it.

Here is my muttrc:
<details>
<summary>Click to expand</summary>
```
source /home/wgm/.mutt/aliases
source /home/wgm/.mutt/keybindings/unbindkeys
source /home/wgm/.mutt/keybindings/keybindings.index
source /home/wgm/.mutt/keybindings/keybindings.attach
source /home/wgm/.mutt/keybindings/keybindings.pager

<!--# set the mailcap file-->
set mailcap_path = ~/.mutt/mailcap
set implicit_autoview = yes

<!--# stop the "press any key to continue"-->
set wait_key=no

<!--# appearance-->
set help=off

<!--# enable threading-->
set sort=threads

<!--# editor-->
set editor=vim

<!--# pager-->
set pager=vim

<!--# delete-->
set delete=yes

<!--# turn on autotagging messages-->
set auto_tag=yes

<!--# format the status bar-->
color status white black
set status_format=''

<!--# disable history-->
set history=0

<!--# use mbsync-->
set mbox_type=Maildir
set folder=~/.mutt/mail
set spoolfile="+INBOX"
unset postponed
unset record
mailboxes +INBOX

<!--# use msmtp-->
set sendmail = "/usr/bin/msmtp"
set realname = "W. Grant Mayer"
set use_from=yes
set envelope_from=yes
set edit_headers = yes
set autoedit

<!--# format dates-->
set index_format=' %<[y?%<[m?%<[d?%[%H:%M ]&%[%a %d]>&%[%b %d]>&%[%m/%y ]>  %S  %-14.14n  %s '

<!--# set autoview up for html-->
auto_view text/html
```
</details>

**Comment by mschilli87** [comment]
Untested, but I'd try something like 

```
text/html; qutebrowser '%s' &
```

in `mailcap` combined with 

```
auto_view text/html
```

in `.muttrc`.

**Comment by wgmayer0** [comment]
Thanks I already have that actually and it still doesn't work

**Comment by elshize** [comment]
@wgmayer0 what exactly do you mean by "automatically" and "any and all"?

Here's what I got in my mailcap:

```
text/html; qutebrowser %s; needsterminal; nametemplate=%s.html
```

and in my `neomutrc`:

```
auto_view text/html
set mailcap_path = ~/.config/neomutt/mailcap
```

When I show the the list of attachments, and press on any that is `text/html`, it opens in qutebrowser. (the nametemplate part is needed specifically for qutebrowser because otherwise it won't interpret the file as html but will print out the raw text)

**Comment by wgmayer0** [comment]
I would like to avoid the attachments step altogether. I would like to
select the message from the index and have it open the html page. I
would also like if I could select multiple emails and qutebrowser could
start loading all attached html documents.

**Comment by flatcap** [comment]
> I would like to avoid the attachments step altogether

That'd be nice, but unfortunately it's a little tricky.

First, when you open the email the attachments don't exist (on disk).
They're part of the file, but they're still encoded.

Second, if you view the HTML part, NeoMutt will save that part as a real file, but only that part.
If the HTML references attached images, then there are two problems.

NeoMutt will need to find and save all the images

The HTML uses references like `cid:1234@5678` rather than filenames,
so NeoMutt would need to translate the references before saving.

---

Mutt was born long before HTML emails were invented.
NeoMutt is playing catchup.

The steps we need to fully support HTML emails aren't that difficult,
but **we need developers** to help with the coding.

I'm happy to mentor any new developers, but I don't have the time to work on this.

**Comment by elshize** [comment]
@wgmayer0 not sure if that helps but I also have a separate binding:

```
macro index,pager B "<view-attachments><search>html<enter><view-mailcap><exit>"
```

It's not exactly what you want but it's something. It works from index & pager. Maybe you can hack it to open html if it finds one or else open message, but I'm not that well versed in neomut macros unfortunately.

Also note that I'm using maildir, and it works fine. From what @flatcap said above, it sounds like it might not work as well with IMAP, at least that's what I understand from that.

**Comment by wgmayer0** [comment]
Yes! That does the trick!

**Comment by wgmayer0** [comment]
This is working perfectly - thanks for the suggestion! I am guessing I
could tag all of those emails and then hit "B" to open all of the html
documents? Thanks

**Comment by dcpurton** [comment]
I'm interested in having a go at this kind of thing.

A possible way forward:

- Create a `related_handler()` function in `handler.c`
- `related_handler()` would need to
    - find all attachments with a `Content-ID` header in the `multipart/related` group
    - save out all attachments in the `multipart/related` group, filtering for `cid:***` for any Content-IDs found
    - display the preferred attachment using the current config logic now that references to related files are fixed up
    - clean up any created files
- A macro could be used to change preferred type to `text/html` and view the email if you normally preferred `text/plain`, but wanted to see the `text/html` part.

Thoughts?

**Comment by wgmayer0** [comment]
Not a bad idea but I lost you at:

> * Create a `related_handler()` function in `handler.c`

Any documentation online that talks about how to do this? Is this messing with the source code before neomutt is compiled?

**Comment by dcpurton** [comment]
> Not a bad idea but I lost you at:
> 
> > * Create a `related_handler()` function in `handler.c`
> 
> Any documentation online that talks about how to do this? Is this messing with the source code before neomutt is compiled?

Yes, I mean adding ability for Neomutt to view the HTML part of a `multipart/related` attachment including any embedded images, which is currently impossible. This requires modifying the Neomutt source. I'm offering to do this :).

I'm pretty sure opening the HTML part automatically should be able to be done using a `message-hook`. The fact that you can't do it yet is probably a configuration problem.

**Comment by wgmayer0** [comment]
Yes this would be a great feature because today, almost all emails that come from organizations are html formatted so this would allow for a quick way to view those emails!

**Comment by wgmayer0** [comment]
> @wgmayer0 not sure if that helps but I also have a separate binding:
> 
> ```
> macro index,pager B "<view-attachments><search>html<enter><view-mailcap><exit>"
> ```
> 
> It's not exactly what you want but it's something. It works from index & pager. Maybe you can hack it to open html if it finds one or else open message, but I'm not that well versed in neomut macros unfortunately.
> 
> Also note that I'm using maildir, and it works fine. From what @flatcap said above, it sounds like it might not work as well with IMAP, at least that's what I understand from that.

Is there a way for this to affect all tagged emails instead of just the selected one?

**Comment by elshize** [comment]
Unfortunately, I have no idea. Maybe @flatcap or @dcpurton can tell if my macro can be tweaked this way.

**Comment by wgmayer0** [comment]
> Unfortunately, I have no idea. Maybe @flatcap or @dcpurton can tell if my macro can be tweaked this way.

Any ideas @flatcap or @dcpurton

**Comment by wgmayer0** [comment]
I have tried everything to get this macro to apply to all tagged messages. I also tried [Special usage: applying to several tagged entries](https://gitlab.com/muttmua/mutt/-/wikis/MuttGuide/Macros) with no luck. I don't really understand how tagging works with macros. It seems like a loop of some sort but I am not sure how to get it to loop through all of the tagged emails and apply `macro index , "<tag-pattern><enter><tag-prefix-cond><view-attachments><search>text/html<enter><view-mailcap><exit><end-cond>"`

**Comment by wgmayer0** [comment]
I want it to open all of the emails in separate qutebrowser windows

**Comment by gahr** [comment]
Some functions are designed to work on a single message and will thus ignore `<tag-prefix>`; `<view-attachments>` is one of these functions.

**Comment by wgmayer0** [comment]
That makes sense thank you
